src/briefcase/integrations/docker.py

```diff
@@ -378,9 +385,406 @@ def check_output(self, args: list[SubprocessArgT], image_tag: str) -> str:
+            # `host.docker.internal` to a network address for the host such that
+            # network-based services on the host can be reached within the container.
+            #
+            # The keyword `host.docker.internal` is already always defined for
+            # containers created by Docker Desktop via the DNS configured within its
+            # Linux VM. However, for Docker Engine, DNS is managed by the Linux host;
+            # therefore, this mapping for `host.docker.internal` must be defined each
+            # time a container is started.
+            #
+            # The keyword `host-gateway` is interpreted and replaced with a network
+            # address for the host by the Docker server when a container starts. The
+            # specific network address used here will be dependent on the Docker
+            # implementation, but it will likely be either the address for `docker0` or
+            # and address for the host otherwise mapped through to the container.
+            subprocess_kwargs.setdefault("add_hosts", []).append(
+                ("host.docker.internal", "host-gateway")
+            )
+
+            # Finally, tell X clients to use the spoofed display
+            subprocess_kwargs.setdefault("env", {}).update(
+                {"DISPLAY": f"host.docker.internal:{proxy_display_num}"}
+            )
+
+            yield subprocess_kwargs
+
+        finally:
+            self.tools.subprocess.cleanup("X display proxy", proxy_popen)
+            xauth_file_path.unlink(missing_ok=True)
+
+    def _x11_tcp_proxy(self, DISPLAY: str) -> tuple[subprocess.Popen, int]:
+        """Starts a TCP proxy as a spoofed X display.
+
+        The TCP server's port is chosen as 6000 + the display number as defined in the
+        X11 standard. In this way, when X clients attempt to make a connection for the
+        chosen display number, this TCP port will be automatically evaluated by the
+        client for consideration as an X display.
+
+        The proxy is bound to 0.0.0.0 and as such will be reachable on all network
+        interfaces defined for the host. Ideally, the proxy would only need to bind to
+        the network bridge defined by Docker, usually `docker0`, but Docker Desktop is
+        not able to use a shared network interface bridge as Docker Engine does since
+        it runs containers inside a Linux VM. Instead, it attaches the host's primary
+        network interface as a device on a virtual network interface bridge created by
+        the VM. Because of this, the proxy must be exposed to the network at large to
+        also be exposed within the container.
+
+        :param DISPLAY: value of DISPLAY environment variable
+        :returns: the Popen process for the proxy and the display number it is using
+        """
+        if not self.tools.shutil.which("socat"):
+            raise BriefcaseCommandError(
+                "Install socat to run an app for a targeted Linux distribution"
```

Comment:
It would be helpful to have distro-specific help here - I was able to guess that `apt install socat` worked on Ubuntu, but that won't necessarily be obvious to everyone, and I don't know if the package name will be the same on other host distros.

Reply:
I'm planning to revisit this with the resolution of https://github.com/beeware/briefcase/issues/1626.

That whole refactor will avoid logging this error so late in the process and will provide an error message relevant to the host system. I just didn't want to do all that in this PR.

You all right if we wait? I'm planning to address it after this is merged.

Reply:
I had the same back-of-my head thought. If #1626 is the next thing on your todo list, I'm fine leaving this a future work.